### ISSUE TITLE
Support dot notation in query projections

### ISSUE THREAD

**Issue by w84thesun** [FerretDB/FerretDB#2430]
### What should be done?

We should add support for dot notation in query projections.

See relevant test cases with this issue number.

The tricky part is that we have to create an empty document when the given path could not be reached.

```js
db.projections.insertMany([
 {
    a: { b: { c: { d: 123 }, e: 13 }, f: 14 },
    g: 15
  },
  {
    a: { b: { c: [ 1, 2 ], e: 13 }, f: 14 },
    g: 15
  },
  {
    a: { b: 12 }
  },
  {
    a: { b: { c: 12 } } 
  }
]);


db.projections.find({}, {'_id': 0,'a.b.c.d': 0})
[
  { a: { b: { c: {}, e: 13 }, f: 14 }, g: 15 },
  {
    a: { b: { c: [ 1, 2 ], e: 13 }, f: 14 },
    g: 15
  },
  { a: { b: 12 } },
  { a: { b: { c: 12 } } }
]


 db.projections.find({}, {'_id': 0,'a.b.c.d': 1})
[
  {
    a: { b: { c: { d: 123 } } }
  },
  { a: { b: { c: [] } } },
  { a: {} },
  { a: { b: {} } }
]

db.projections.find({}, {'_id': 0,'a.b.c': 0})
[
  { a: { b: { e: 13 }, f: 14 }, g: 15 },
  { a: { b: { e: 13 }, f: 14 }, g: 15 },
  { a: { b: 12 } },
  { a: { b: {} } }
]


db.projections.find({}, {'_id': 0,'a.b.c': 1})
[
  {
    a: { b: { c: { d: 123 } } }
  },
  {
    a: { b: { c: [ 1, 2 ] } }
  },
  { a: {} },
  { a: { b: { c: 12 } } }
]

```

### Where?

internal/handlers/common/projection.go

### Definition of Done

- all handlers updated;
- unit tests added/updated;
- integration/compatibility tests added/updated;
- spot refactorings done;
- user documentation updated;
- something else?

**Comment by AlekSi** [comment]
https://github.com/FerretDB/FerretDB/pull/2631/files#r1194617322

It is also not clear for me why this issue is Done

**Comment by chilagrow** [comment]
> https://github.com/FerretDB/FerretDB/pull/2631/files#r1194617322
> 
> It is also not clear for me why this issue is Done

:wave: @AlekSi what part of of this issue it not done?

**Comment by AlekSi** [comment]
There is only one PR that separates codebases. I don't see how separating codebases would solve that issue.

There are no other PRs with tests, and the issue descriptions specifically talks about them.

**Comment by chilagrow** [comment]
> There is only one PR that separates codebases. I don't see how separating codebases would solve that issue.
> 
> There are no other PRs with tests, and the issue descriptions specifically talks about them.

I think I misunderstood perhaps, the issue title is specific for query projection, so I got understanding that it will not have aggregation projection. So far I created issue for handling aggregation projection in https://github.com/FerretDB/FerretDB/issues/2633. But happy to do it part of this issue, after all I created separate code base for aggregation projection :hugs:

**Comment by AlekSi** [comment]
> See relevant test cases with this issue number.

Do we have them?

**Comment by chilagrow** [comment]
> > See relevant test cases with this issue number.
> 
> Do we have them?

Where is that from? No we don't have relevant test cases, we have highlighted un-tested path in the code base for #2633.

**Comment by AlekSi** [comment]
It is a quite from the issue description. It also contains examples that are not being added as tests in the PR that supposedly only makes a copy of existing codebase.

**Comment by chilagrow** [comment]
> It is a quote from the issue description. It also contains examples that are not being added as tests in the PR that supposedly only makes a copy of existing codebase.

My bad :see_no_evil: I was being blind. Yes we did have those for query projection, unfortunately I removed them in this PR as I thought I covered them all. https://github.com/FerretDB/FerretDB/commit/2516bb192eb872f8466f4acd558bf6391dff4db7#diff-814d467d6be43ec152f53e8faa0019d312710efbd798d0558f597424fa72f507

And we can apply the same test cases for aggregation projection.
